### PR TITLE
MWPW-133754 - Support structured data API for SEO

### PR DESCRIPTION
**Pull request by hparra**
Add support for `seotech-structured-data` feature. We refactor existing support for `seotech-video-url` and add new tests and make improvements to existing ones.

Resolves: [MWPW-133754](https://jira.corp.adobe.com/browse/MWPW-133754)

**Test URLs:**

- Before: https://milo.adobe.com/drafts/hgpa/test/seotech-structured-data/structured-data?martech=off&seotech-sheet-url=https://milo.adobe.com/drafts/hgpa/test/seotech-structured-data/structured-data.json
- After: https://seotech-structured-data--milo--hparra.hlx.live/drafts/hgpa/test/seotech-structured-data/structured-data?martech=off&seotech-sheet-url=https://milo.adobe.com/drafts/hgpa/test/seotech-structured-data/structured-data.json

**Validation:**

The following should return a `FAQPage` JSON-LD object:

```js
JSON.parse(document.querySelector('script[type="application/ld+json"]').text)
```

**References:**

- https://github.com/adobecom/milo/pull/1065

### CODE REVIEW

File: libs/features/seotech/seotech.js [hunk 1]
```diff
@@ -21,14 +21,45 @@ export async function getVideoObject(url, seotechAPIUrl) {
...
+  const apiUrl = new URL(seotechAPIUrl);
+  apiUrl.pathname = '/api/v1/web/seotech/getStructuredData';
+  apiUrl.searchParams.set('url', url);
+  if (sheetUrl) {
+    apiUrl.searchParams.set('sheetUrl', sheetUrl);
+  }
+  const resp = await fetch(apiUrl.href, { headers: { 'Content-Type': 'application/json' } });
+  const body = await resp?.json();
+  if (!resp.ok) {
+    throw new Error(`Failed to fetch structured data: ${body?.error}`);
```
Comment (overmyheadandbody) [inline libs/features/seotech/seotech.js:34]: Should this also contain the `sheetUrl` so the Splunk messages are clear as to which sheet failed?

File: libs/features/seotech/seotech.js [hunk 2]
```diff
@@ -21,14 +21,45 @@ export async function getVideoObject(url, seotechAPIUrl) {
...
+    prs.push(getStructuredData(pageUrl, sheetUrl, seotechAPIUrl)
+      .then((r) => r.forEach((obj) => append(obj)))
+      .catch((e) => logError(e.message)));
   }
+  if (getMetadata('seotech-video-url')) {
+    prs.push(getVideoObject(getMetadata('seotech-video-url'), seotechAPIUrl)
+      .then((r) => append(r))
+      .catch((e) => logError(e.message)));
+  }
+  return Promise.all(prs);
```
Comment (overmyheadandbody) [inline libs/features/seotech/seotech.js:62]: Is there a need to further chain this in the calling function?

**Review by overmyheadandbody — APPROVED** [review]

**Review by chrischrischris — APPROVED** [review]

File: libs/features/seotech/seotech.js (hunk 1 above)
Comment (hparra) [reply libs/features/seotech/seotech.js:34]: `sheetUrl` is used by the service, so if there is a problem with it then the URL should be returned in `body.error` msg. Will fix api to do this.

File: libs/features/seotech/seotech.js (hunk 2 above)
Comment (hparra) [reply libs/features/seotech/seotech.js:62]: Yes? Caller is `then` of another promise in impl. We also `await` in tests.

File: libs/features/seotech/README.md [hunk 3]
```diff
@@ -19,3 +19,17 @@ Video Platforms:
...
+
+Metadata Properties:
+
+- `seotech-structured-data`: `on` to enable SEOTECH lookup
+- `seotech-sheet-url`: url of Franklin Spreadsheet JSON (Optional)
+
+You can also specify `seotech-sheet-url` as a query parameter.
+Otherwise SEOTECH will search for _/structured-data.json_ at the root of the current page.
+
+Please see For more details see [SEOTECH](https://wiki.corp.adobe.com/display/seoteam/SEOTECH+API) (Corp Only) for list of supported structured data.
```
[the remark this answers is not in the document]
Comment (hparra) [reply libs/features/seotech/README.md]: https://github.com/adobecom/milo/pull/1115/commits/c8a055760fba56c79062bb55009e99e2ef99bcdf

File: libs/features/seotech/seotech.js [hunk 4]
```diff
@@ -21,14 +21,45 @@ export async function getVideoObject(url, seotechAPIUrl) {
...
+  }
+  const resp = await fetch(apiUrl.href, { headers: { 'Content-Type': 'application/json' } });
+  const body = await resp?.json();
+  if (!resp.ok) {
+    throw new Error(`Failed to fetch structured data: ${body?.error}`);
+  }
+  return body.objects;
+}
+
+export async function appendScriptTag({ locationUrl, getMetadata, createTag, getConfig }) {
```
Comment (mokimo) [inline libs/features/seotech/seotech.js:39]: Why not just import `getMetadata, createTag, getConfig`

File: libs/features/seotech/seotech.js [hunk 5]
```diff
@@ -21,14 +21,45 @@ export async function getVideoObject(url, seotechAPIUrl) {
...
-    const obj = await getVideoObject(url, seotechAPIUrl);
+
+  const append = (obj) => {
     const script = createTag('script', { type: 'application/ld+json' }, JSON.stringify(obj));
     document.head.append(script);
-  } catch (e) {
-    logError(e.message);
+  };
+
+  const prs = [];
```
Comment (mokimo) [inline libs/features/seotech/seotech.js:49]: prs?

**Review by mokimo — APPROVED** [review]

File: libs/features/seotech/seotech.js (hunk 5 above)
Comment (overmyheadandbody) [reply libs/features/seotech/seotech.js:49]: _Promises_, I guess

Comment (hparra) [reply libs/features/seotech/seotech.js:49]: Yes. Does this repo prefer I spell these  vars out?

File: libs/features/seotech/seotech.js (hunk 4 above)
Comment (hparra) [reply libs/features/seotech/seotech.js:39]: This is convention of similar features as it prevents import cycle. It also makes testing a little easier

File: libs/features/seotech/seotech.js (hunk 5 above)
Comment (hparra) [reply libs/features/seotech/seotech.js]: https://github.com/adobecom/milo/pull/1115/commits/40dc3fec84d0266e4294182fd506011b58e80fdb

**Pull request merged**